src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetLockService.cs

```diff
@@ -22,6 +22,28 @@ public sealed class NuGetLockService : INuGetLockService, IDisposable
 
         public bool IsLockHeld => _semaphore.CurrentCount == 0;
 
+        public async Task WaitAndReleaseAsync(CancellationToken token)
+        {
+            if (_lockCount.Value == 0)
+            {
+                _lockCount.Value++;
+                try
+                {
+                    await _semaphore.WaitAsync(token);
+                }
+                finally
+                {
+                    try
+                    {
+                        _semaphore.Release();
```

Comment:
I don't think I follow it. But how this thread release will release the semaphore held with another thread? Wouldn't that be a noop if thread didn't acquire the lock?

We can also discuss it in person.